apm/event.go

```diff
@@ -47,6 +47,7 @@ func (e *eventWindow) Message() message.Composer {
 	}
 	colls := make([]output, 0, len(e.data))
 	for k, v := range e.data {
+		v.mutex.RLock()
```

Comment:
Is there a reason you chose to acquire the read lock inside the for loop? My intuition is that this could cause the for loop to hang, whereas acquiring it immediately before entering the loop ensures that the loop completes. I haven't read the code close enough to justify my intuition though.

Reply:
v is a variable defined only inside the scope of the for loop so it needs to be locked inside the loop. The data we're trying to protect is the Succeeded/Failed counters on eventRecord, which might be updated by the mongo driver asynchronously.

`e.data` does not need a lock itself. It is initialized once in `basicMonitor.Rotate()` and never written to again. The underlying `current` map is immediately set to a new empty map, ensuring that `e.data` does not receive more writes.

The things that are not thread-safe in the monitor are the underlying counters in the eventRecord (`eventRecord.Succeeded` and `eventRecord.Failed`). 
These are updated by the functions defined in `*basicMonitor.DriverAPM()`, which are called by the mongo driver asynchronously to write to the event counters for a given namespace/db/collection. Assuming the lock is fair, this flushing goroutine will eventually acquire the lock once the mongo driver is not writing actively to the counters.